freqtrade/strategy/backtest_lookahead_bias_checker.py

```diff
@@ -0,0 +1,252 @@
+        self.backtesting._set_strategy(self.backtesting.strategylist[0])
+        varHolder.data, varHolder.timerange = self.backtesting.load_bt_data()
+        self.backtesting.load_bt_data_detail()
+
+        varHolder.indicators = self.backtesting.strategy.advise_all_indicators(varHolder.data)
+        varHolder.result = self.get_result(self.backtesting, varHolder.indicators)
+
+    def start(self, config, strategy_obj: dict, args) -> None:
+
+        # deepcopy so we can change the pairs for the 2ndary runs
+        # and not worry about another strategy to check after.
+        self.local_config = deepcopy(config)
+        self.local_config['strategy_list'] = [strategy_obj['name']]
+        self.current_analysis = Analysis()
+        self.minimum_trade_amount = args['minimum_trade_amount']
+        self.targeted_trade_amount = args['targeted_trade_amount']
+        self.exportfilename = args['exportfilename']
+        self.strategy_obj = strategy_obj
+
+        # first make a single backtest
+        self.full_varHolder = VarHolder()
+
+        # define datetime in human-readable format
+        parsed_timerange = TimeRange.parse_timerange(config['timerange'])
+
+        if parsed_timerange.startdt is None:
+            self.full_varHolder.from_dt = datetime.utcfromtimestamp(0)
+        else:
+            self.full_varHolder.from_dt = parsed_timerange.startdt
+
+        if parsed_timerange.stopdt is None:
+            self.full_varHolder.to_dt = datetime.now()
+        else:
+            self.full_varHolder.to_dt = parsed_timerange.stopdt
+
+        self.prepare_data(self.full_varHolder, self.local_config['pairs'])
+
+        found_signals: int = self.full_varHolder.result['results'].shape[0] + 1
+        if found_signals >= self.targeted_trade_amount:
+            print(f"Found {found_signals} trades, calculating {self.targeted_trade_amount} trades.")
+        elif self.targeted_trade_amount >= found_signals >= self.minimum_trade_amount:
+            print(f"Only found {found_signals} trades. Calculating all available trades.")
+        else:
+            print(f"found {found_signals} trades "
+                  f"which is less than minimum_trade_amount {self.minimum_trade_amount}. "
+                  f"Cancelling this backtest lookahead bias test.")
+            return
+
+        # now we loop through all entry signals
+        # starting from the same datetime to avoid miss-reports of bias
+        for idx, result_row in self.full_varHolder.result['results'].iterrows():
+            if self.current_analysis.total_signals == self.targeted_trade_amount:
+                break
+
+            # if force-sold, ignore this signal since here it will unconditionally exit.
+            if result_row.close_date == self.dt_to_timestamp(self.full_varHolder.to_dt):
+                continue
+
+            self.current_analysis.total_signals += 1
+
+            self.entry_varHolder = VarHolder()
+            self.exit_varHolder = VarHolder()
```

Comment:
You may want to have these as local variables instead. That'll make their actual scope clear (they are only valid within the loop).

Reply:
Yes, moved the self.entry_varHolder and exit too only as self-reference of the self.entry_varHolders[idx]
The for loop will care for picking of which varHolders to use

Left         self.current_analysis.total_signals += 1 in place so even if something fails it reports the right lines out without any additional lines (it s just a +=1 so it really doesnt take any significant computational power even though when everything works ok it is some wasted cycles)